ast-visual-studio-extension/CxExtension/Services/ASCAService.cs

```diff
@@ -0,0 +1,490 @@
+﻿using ast_visual_studio_extension.CxWrapper.Models;
+using EnvDTE;
+using EnvDTE80;
+using Microsoft.VisualStudio.Shell;
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading.Tasks;
+using System.Timers;
+using System.IO;
+
+
+
+using Microsoft.VisualStudio.TextManager.Interop;
+using MARKERTYPE = Microsoft.VisualStudio.TextManager.Interop.MARKERTYPE;
+using Microsoft.VisualStudio;
+
+namespace ast_visual_studio_extension.CxExtension.Services
+{
+    public class ASCAService
+    {
+        private readonly System.Timers.Timer _debounceTimer;
+
+        private readonly CxCLI.CxWrapper cxWrapper;
+        private const int DEBOUNCE_DELAY = 2000; // Delay of 2 seconds
+        private bool _isSubscribed = false;
+        private bool _isInitialized = false;
+
+        private readonly DTE2 _dte;
+        private string _lastDocumentContent = string.Empty;
+        private OutputWindowPane _outputPane;
+        private static ASCAService _instance;
+        private static readonly object _lock = new object();
+        private TextEditorEvents _textEditorEvents;
+        private ErrorListProvider _errorListProvider;
+        private List<IVsTextLineMarker> _activeMarkers = new List<IVsTextLineMarker>();
+
+
+
+        private ASCAService(CxCLI.CxWrapper cxWrapper)
+        {
+            this.cxWrapper = cxWrapper;
+            _debounceTimer = new System.Timers.Timer(DEBOUNCE_DELAY);
+            _debounceTimer.Elapsed += OnDebounceTimerElapsed;
+            _debounceTimer.AutoReset = false;
+
+            _dte = Package.GetGlobalService(typeof(DTE)) as DTE2;
+            InitializeOutputPane();
+        }
+
+        public static ASCAService GetInstance(CxCLI.CxWrapper cxWrapper)
+        {
+            if (_instance == null)
+            {
+                lock (_lock)
+                {
+                    if (_instance == null)
+                    {
+                        _instance = new ASCAService(cxWrapper);
+                    }
+                }
+            }
+            return _instance;
+        }
+
+        private void InitializeOutputPane()
+        {
+            ThreadHelper.ThrowIfNotOnUIThread();
+
+            var outputWindow = _dte.ToolWindows.OutputWindow;
+            _outputPane = outputWindow.OutputWindowPanes
+                .Cast<OutputWindowPane>()
+                .FirstOrDefault(p => p.Name == "Checkmarx") ?? outputWindow.OutputWindowPanes.Add("Checkmarx");
+            _outputPane.Activate();
+        }
+
+        private void WriteToOutputPane(string message)
```

Comment:
maybe better to create a new util class to handle all the output actions

Reply:
At this stage, I preferred not to create a new class for this, since it’s just one function and is currently used only by ASCA. However, in the future, if we need it for additional cases, we will create a new utility class to handle all the output actions.